File: rental/src/main/java/io/axoniq/demo/bikerental/rental/ui/RentalController.java
package io.axoniq.demo.bikerental.rental.ui;

import io.axoniq.demo.bikerental.coreapi.payment.ConfirmPaymentCommand;
import io.axoniq.demo.bikerental.coreapi.payment.PaymentStatus;
import io.axoniq.demo.bikerental.coreapi.payment.PaymentStatusNamedQueries;
import io.axoniq.demo.bikerental.coreapi.rental.BikeStatus;
import io.axoniq.demo.bikerental.coreapi.rental.BikeStatusNamedQueries;
import io.axoniq.demo.bikerental.coreapi.rental.RegisterBikeCommand;
import io.axoniq.demo.bikerental.coreapi.rental.RequestBikeCommand;
import io.axoniq.demo.bikerental.coreapi.rental.ReturnBikeCommand;
import io.axoniq.demo.bikerental.rental.support.CommandDispatcher;
import io.axoniq.demo.bikerental.rental.support.QueryDispatcher;
import org.springframework.http.codec.ServerSentEvent;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.List;
import java.util.Objects;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

@CrossOrigin(origins = "*", maxAge = 3600)
@RestController
@RequestMapping("/")
public class RentalController {

    private final CommandDispatcher commandDispatcher;
    private final QueryDispatcher queryDispatcher;
    private final BikeRentalDataGenerator bikeRentalDataGenerator;

    public RentalController(CommandDispatcher commandDispatcher,
                            QueryDispatcher queryDispatcher,
                            BikeRentalDataGenerator bikeRentalDataGenerator) {
        this.commandDispatcher = commandDispatcher;
        this.queryDispatcher = queryDispatcher;
        this.bikeRentalDataGenerator = bikeRentalDataGenerator;
    }

    @PostMapping("/bikes")
    public CompletableFuture<String> registerBike(@RequestParam("bikeType") String bikeType,
                                                  @RequestParam("location") String location) {
        String bikeId = UUID.randomUUID().toString();
        return commandDispatcher.send(new RegisterBikeCommand(bikeId, bikeType, location))
                                .thenApply(unused -> bikeId);
    }

    @PostMapping("/bikes/batch")
    public CompletableFuture<Void> generateBikes(@RequestParam("count") int bikeCount,
                                                 @RequestParam(value = "type") String bikeType) {
        return bikeRentalDataGenerator.generateBikes(bikeCount, bikeType);
    }

    @GetMapping("/bikes")
    public CompletableFuture<List<BikeStatus>> findAll() {
        return queryDispatcher.queryMany(BikeStatusNamedQueries.FIND_ALL, null, BikeStatus.class);
    }

    @GetMapping("/bikeUpdates")
    public Flux<ServerSentEvent<String>> subscribeToAllUpdates() {
        return queryDispatcher.subscriptionQueryMany(BikeStatusNamedQueries.FIND_ALL, null, BikeStatus.class)
                              .map(BikeStatus::description)
                              .map(description -> ServerSentEvent.builder(description).build());
    }

    @GetMapping("/bikeUpdatesJson")
    public Flux<ServerSentEvent<BikeStatus>> subscribeToAllUpdatesJson() {
        return queryDispatcher.subscriptionQueryMany(BikeStatusNamedQueries.FIND_ALL, null, BikeStatus.class)
                              .map(status -> ServerSentEvent.builder(status).build());
    }

    @GetMapping("/bikeUpdates/{bikeId}")
    public Flux<ServerSentEvent<String>> subscribeToBikeUpdates(@PathVariable("bikeId") String bikeId) {
        return queryDispatcher.subscriptionQuery(BikeStatusNamedQueries.FIND_ONE, bikeId, BikeStatus.class)
                              .map(BikeStatus::description)
                              .map(description -> ServerSentEvent.builder(description).build());
    }

    @PostMapping("/requestBike")
    public CompletableFuture<String> requestBike(@RequestParam("bikeId") String bikeId,
                                                 @RequestParam(value = "renter", required = false) String renter) {
        return commandDispatcher.send(
                new RequestBikeCommand(bikeId, renter != null ? renter : bikeRentalDataGenerator.randomRenter()),
                String.class);
    }

    @PostMapping("/returnBike")
    public CompletableFuture<Void> returnBike(@RequestParam("bikeId") String bikeId) {
        return commandDispatcher.send(new ReturnBikeCommand(bikeId, bikeRentalDataGenerator.randomLocation()));
    }

    @GetMapping("findPayment")
    public Mono<String> getPaymentId(@RequestParam("reference") String paymentRef) {
        return queryDispatcher.subscriptionQuery(PaymentStatusNamedQueries.GET_PAYMENT_ID, paymentRef, String.class)
                              .filter(Objects::nonNull)
                              .next();
    }

    @GetMapping("pendingPayments")
    public CompletableFuture<List<PaymentStatus>> getPendingPayments() {
        return queryDispatcher.queryMany(PaymentStatusNamedQueries.GET_ALL_PAYMENTS,
                                         PaymentStatus.Status.PENDING, PaymentStatus.class);
    }

    @PostMapping("acceptPayment")
    public CompletableFuture<Void> acceptPayment(@RequestParam("id") String paymentId) {
        return commandDispatcher.send(new ConfirmPaymentCommand(paymentId));
    }

    @GetMapping(value = "watch", produces = "text/event-stream")
    public Flux<String> watchAll() {
        return queryDispatcher.subscriptionQueryMany(BikeStatusNamedQueries.FIND_ALL, null, BikeStatus.class)
                              .map(bs -> bs.getBikeId() + " -> " + bs.description());
    }

    @GetMapping(value = "watch/{bikeId}", produces = "text/event-stream")
    public Flux<String> watchBike(@PathVariable("bikeId") String bikeId) {
        return queryDispatcher.subscriptionQuery(BikeStatusNamedQueries.FIND_ONE, bikeId, BikeStatus.class)
                              .map(bs -> bs.getBikeId() + " -> " + bs.description());
    }

    @PostMapping(value = "/generateRentals")
    public Flux<String> generateData(@RequestParam(value = "bikeType") String bikeType,
                                     @RequestParam("loops") int loops,
                                     @RequestParam(value = "concurrency", defaultValue = "1") int concurrency,
                                     @RequestParam(value = "abandonPaymentFactor", defaultValue = "100") int abandonPaymentFactor,
                                     @RequestParam(value = "delay", defaultValue = "0") int delay) {
        return bikeRentalDataGenerator.generateRentals(bikeType, loops, concurrency, abandonPaymentFactor, delay);
    }

    @GetMapping("/bikes/{bikeId}")
    public CompletableFuture<BikeStatus> findStatus(@PathVariable("bikeId") String bikeId) {
        return queryDispatcher.query(BikeStatusNamedQueries.FIND_ONE, bikeId, BikeStatus.class);
    }
}

File: rental/src/ui/rental-controller.ts
import { randomUUID } from 'node:crypto';
import { Router, type Response } from 'express';
import {
  BikeStatus,
  BikeStatusNamedQueries,
  CommandDispatcher,
  ConfirmPaymentCommand,
  JavaTypes,
  PaymentStatus,
  PaymentStatusNamedQueries,
  PaymentStatusValue,
  QueryDispatcher,
  RegisterBikeCommand,
  RequestBikeCommand,
  ReturnBikeCommand,
} from '@bikerental/core-api';
import { BikeStatusProjection } from '../query/bike-status-projection.js';
import { BikeRentalDataGenerator } from './bike-rental-data-generator.js';

function openSse(res: Response): void {
  res.setHeader('Content-Type', 'text/event-stream');
  res.setHeader('Cache-Control', 'no-cache');
  res.setHeader('Connection', 'keep-alive');
  res.flushHeaders();
}

function str(value: unknown): string {
  return value === undefined ? '' : String(value);
}

export function rentalRouter(
  commandDispatcher: CommandDispatcher,
  queryDispatcher: QueryDispatcher,
  projection: BikeStatusProjection,
  dataGenerator: BikeRentalDataGenerator,
): Router {
  const router = Router();

  router.post('/bikes', async (req, res) => {
    const bikeId = randomUUID();
    await commandDispatcher.send(
      new RegisterBikeCommand(bikeId, str(req.query.bikeType), str(req.query.location)),
    );
    res.type('text/plain').send(bikeId);
  });

  router.post('/bikes/batch', async (req, res) => {
    await dataGenerator.generateBikes(Number(req.query.count), str(req.query.type));
    res.status(200).end();
  });

  router.get('/bikes', async (_req, res) => {
    res.json(await queryDispatcher.queryMany<BikeStatus>(BikeStatusNamedQueries.FIND_ALL, null, BikeStatus.TYPE));
  });

  router.get('/bikeUpdates', (req, res) => {
    openSse(res);
    const unsubscribe = projection.subscribeAll((status) => res.write(`data:${status.description()}\n\n`));
    req.on('close', unsubscribe);
  });

  router.get('/bikeUpdatesJson', (req, res) => {
    openSse(res);
    const unsubscribe = projection.subscribeAll((status) => res.write(`data:${JSON.stringify(status)}\n\n`));
    req.on('close', unsubscribe);
  });

  router.get('/bikeUpdates/:bikeId', (req, res) => {
    openSse(res);
    const unsubscribe = projection.subscribeOne(req.params.bikeId, (status) =>
      res.write(`data:${status.description()}\n\n`),
    );
    req.on('close', unsubscribe);
  });

  router.post('/requestBike', async (req, res) => {
    const renter = req.query.renter ? str(req.query.renter) : dataGenerator.randomRenter();
    const paymentRef = await commandDispatcher.send<string>(
      new RequestBikeCommand(str(req.query.bikeId), renter),
      JavaTypes.STRING,
    );
    res.type('text/plain').send(paymentRef ?? '');
  });

  router.post('/returnBike', async (req, res) => {
    await commandDispatcher.send(new ReturnBikeCommand(str(req.query.bikeId), dataGenerator.randomLocation()));
    res.status(200).end();
  });

  router.get('/findPayment', async (req, res) => {
    const paymentId = await queryDispatcher.pollFor<string>(
      PaymentStatusNamedQueries.GET_PAYMENT_ID,
      str(req.query.reference),
      JavaTypes.STRING,
      (result) => result !== null,
    );
    res.type('text/plain').send(paymentId);
  });

  router.get('/pendingPayments', async (_req, res) => {
    res.json(
      await queryDispatcher.queryMany<PaymentStatus>(
        PaymentStatusNamedQueries.GET_ALL_PAYMENTS,
        PaymentStatusValue.PENDING,
        PaymentStatus.TYPE,
      ),
    );
  });

  router.post('/acceptPayment', async (req, res) => {
    await commandDispatcher.send(new ConfirmPaymentCommand(str(req.query.id)));
    res.status(200).end();
  });

  router.get('/watch', (req, res) => {
    openSse(res);
    const unsubscribe = projection.subscribeAll((bs) =>
      res.write(`data:${bs.getBikeId()} -> ${bs.description()}\n\n`),
    );
    req.on('close', unsubscribe);
  });

  router.get('/watch/:bikeId', (req, res) => {
    openSse(res);
    const unsubscribe = projection.subscribeOne(req.params.bikeId, (bs) =>
      res.write(`data:${bs.getBikeId()} -> ${bs.description()}\n\n`),
    );
    req.on('close', unsubscribe);
  });

  router.post('/generateRentals', async (req, res) => {
    res.type('text/plain');
    await dataGenerator.generateRentals(
      str(req.query.bikeType),
      Number(req.query.loops),
      req.query.concurrency ? Number(req.query.concurrency) : 1,
      req.query.abandonPaymentFactor ? Number(req.query.abandonPaymentFactor) : 100,
      req.query.delay ? Number(req.query.delay) : 0,
      (line) => res.write(line),
    );
    res.end();
  });

  router.get('/bikes/:bikeId', async (req, res) => {
    res.json(await queryDispatcher.query<BikeStatus>(BikeStatusNamedQueries.FIND_ONE, req.params.bikeId, BikeStatus.TYPE));
  });

  return router;
}
